Add CI and Fastlane automation for one-off and nightly TestFlight builds

mokagio opened the pull request:
What is says in the title, but notice that a few code and `Info.plist` changes were required to pass the TestFlight validation.

Here's a screenshot from a successful upload...

<img width="910" alt="image" src="https://github.com/user-attachments/assets/52b0d540-5b5a-4317-bdcc-149228c5bee1" />

...which I was able to download via the iOS TestFlight app...

<img alt="IMG_1CA11FF065C4-1" src="https://github.com/user-attachments/assets/825aa6e5-0d6a-4002-ad4e-f908e373d18e" width=300/>

...but which crashed on launch with

```
DYLD error: Library not loaded: @rpath/BuildSettingsKit.framework/BuildSettingsKit
```

I think the crash is unrelated with the infra we put in place here. I'd like to merge this PR ASAP and happy to follow up.

mokagio committed:
Explicitly set .ipa output name when building for ASC

We'll use it to derive the path when uploading the ipa to TestFlight in
a dedicated lane.

mokagio committed:
Add lane to upload Reader build to TestFlight

mokagio committed:
Hack a version long and short to use for TestFlight upload testing

mokagio committed:
Add dedicated pipeline to upload Reader to TestFlight

mokagio committed:
Add trigger button to send Reader build to TestFlight

dangermattic commented on fastlane/lanes/build.rb:
```diff
@@ -277,11 +286,31 @@
 
+  lane :upload_to_app_store_connect_reader do
+    # Eventually, this will be replaced with a real release notes file. Or maybe not.
+    temp_release_notes = File.join(Dir.tmpdir, "reader_release_notes.md")
```
<table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/wordpress-mobile/WordPress-iOS/blob/ef70bfc42fd7c3bdd398cc1b532c81d5d4c807f8/fastlane/lanes/build.rb#L298"></span>Style/StringLiterals: Prefer single-quoted strings when you don't need string interpolation or special symbols.</td>
    </tr>
  </tbody>
</table>

mokagio replied:
_Cursor wrote this_ is not an excuse for not noticing the style violation locally.

There must be something off in my RuboCop-Cursor integration...

dangermattic commented on fastlane/lanes/build.rb:
```diff
@@ -277,11 +286,31 @@
+    temp_release_notes = File.join(Dir.tmpdir, "reader_release_notes.md")
+
+    begin
+      File.write(temp_release_notes, "Thank you for testing the new Reader app. Please get in touch with any feedback or suggestions.")
```
<table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/wordpress-mobile/WordPress-iOS/blob/ef70bfc42fd7c3bdd398cc1b532c81d5d4c807f8/fastlane/lanes/build.rb#L301"></span>Style/StringLiterals: Prefer single-quoted strings when you don't need string interpolation or special symbols.</td>
    </tr>
  </tbody>
</table>

dangermattic commented on fastlane/lanes/build.rb:
```diff
@@ -277,11 +286,31 @@
+        beta_app_description_path: BETA_APP_DESCRIPTION_PATH_READER
+      )
+    ensure
+      File.delete(temp_release_notes) if File.exist?(temp_release_notes)
```
<table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/wordpress-mobile/WordPress-iOS/blob/ef70bfc42fd7c3bdd398cc1b532c81d5d4c807f8/fastlane/lanes/build.rb#L310"></span>Lint/NonAtomicFileOperation: Use atomic file operation method <code>FileUtils.rm_f</code>.</td>
    </tr>
  </tbody>
</table>

dangermattic replied:
<table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/wordpress-mobile/WordPress-iOS/blob/ef70bfc42fd7c3bdd398cc1b532c81d5d4c807f8/fastlane/lanes/build.rb#L310"></span>Lint/NonAtomicFileOperation: Remove unnecessary existence check <code>File.exist?</code>.</td>
    </tr>
  </tbody>
</table>

dangermattic commented:
<!--
  0 Errors
  2 Warnings: <code>Modules/Package.swift</c...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><code>Modules/Package.swift</code> was changed without updating its corresponding <code>Package.resolved</code>. Please resolve the Swift packages as appropriate to your project setup (e.g. in Xcode or by running <code>swift package resolve</code>).</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/wordpress-mobile/WordPress-iOS/milestone/308">25.9</a>. The due date for this milestone has already passed.<br />
Please assign it to a milestone with a later deadline or check whether the release for this milestone has already been finished.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

mokagio committed:
Address RuboCop violations

mokagio committed:
Only install gems in Reader upload script

No need to set up Swift packages or XCFrameworks.

mokagio committed:
Set `VERSION_LONG` as `BUILDKITE_BUILD_NUMBER` default

mokagio committed:
Specify supported device orientation for iPhone and iPad

Aims to address:

 [Application Loader Error Output]: [ContentDelivery.Uploader.6000033F8180] Validation failed (409) Invalid bundle. No orientations were specified in the com.automattic.reader bundle. To support iPad multitasking, specify the “UIInterfaceOrientationPortrait,UIInterfaceOrientationPortraitUpsideDown,UIInterfaceOrientationLandscapeLeft,UIInterfaceOrientationLandscapeRight” orientations for the UISupportedInterfaceOrientations Info.plist key. For details, visit: https://developer.apple.com/documentation/bundleresources/information_property_list/uisupportedinterfaceorientations. (ID: b7c3b053-0ea8-4267-83dc-1dee7ea62885)

See
https://buildkite.com/automattic/wordpress-ios/builds/27266/steps?sid=0196236d-400e-4be1-9cf8-399f2ca6b45c

mokagio commented on .buildkite/pipeline.yml:
```diff
@@ -6,17 +6,35 @@ env:
-    notify:
-      - github_commit_status:
-          context: "Reader ASC Demo Build"
+      - block: Distribute Reader via TestFlight
+        prompt: Distribute the Reader build via TestFlight?
+        key: testflight_triggered_reader
+        # Block steps have implicit dependency on the steps that come before them.
+        # See https://buildkite.com/docs/pipelines/trigger-step
+        #
+        # Make it depend on nothing so we don't have to wait for previous steps to finish before the deployment can start.
+        depends_on: ~
```
<img width="978" alt="image" src="https://github.com/user-attachments/assets/98012ccc-234e-4d37-b159-3c8b9e0dae9c" />

mokagio committed:
Make Reader upload depend on both build and block step

I was expecting the implicit dependency in the block to force the upload
to wait till the build finished, but that was not the case in
https://buildkite.com/automattic/wordpress-ios/builds/27273/steps?sid=019624d9-64df-4b64-89da-927be7ecb550

mokagio commented on .buildkite/commands/prototype-build-jetpack.sh:
```diff
@@ -1,14 +1,7 @@
-brew tap FelixHerrmann/tap
-brew install swift-package-list
-
 "$(dirname "${BASH_SOURCE[0]}")/shared-set-up.sh"
+"$(dirname "${BASH_SOURCE[0]}")/shared-set-up-distribution.sh"
```
All the deleted code went into `shared-set-up-distribution.sh`.

mokagio commented on .buildkite/commands/shared-set-up-distribution.sh:
```diff
@@ -0,0 +1,11 @@
+brew tap FelixHerrmann/tap
+brew install swift-package-list
+
+brew install imagemagick
+brew install ghostscript
```
I need to double check, but I think we can remove these. They would have been used to updated the app icon in the installable builds, but I couldn't find the code for it when I looked.

mokagio commented on .buildkite/pipeline.yml:
```diff
@@ -6,6 +6,35 @@ env:
+        depends_on: ~
+
+      - label: ":eyeglasses: Reader one-off TestFlight Upload"
+        depends_on: [build_asc_reader, testflight_triggered_reader]
```
I was expecting `testflight_triggered_reader` to suffice because of the implicit before-after dependency in `block`, but that's not what happened in https://buildkite.com/automattic/wordpress-ios/builds/27273/steps?sid=019624d9-64df-4b64-89da-927be7ecb550

While it's possible I made a mistake elsewhere, this verbose explicit syntax works so I think we can keep it for now.

mokagio commented:
@wordpress-mobile/apps-infrastructure this is still a draft because at the time of writing ASC seems stuck pre-processing the build (may because it's the first uploaded?)

![image](https://github.com/user-attachments/assets/1676a968-c09f-47a9-8012-d665071a1a82)

But I'd still appreciated a look in the meantime. Thanks!

mokagio commented on fastlane/lanes/build.rb:
```diff
@@ -266,7 +275,9 @@
     )
   end
 
-  lane :build_for_app_store_connect_reader do
+  lane :build_for_app_store_connect_reader do |build_number: ENV.fetch('BUILDKITE_BUILD_NUMBER', nil)|
+    UI.user_error!('No build number provided and BUILDKITE_BUILD_NUMBER environment variable is not set') if build_number.nil?
```
Either before merging or in a followup, I think it would be better to remove the default `BUILDKITE_BUILD_NUMBER` value in favor of having the CI script passing it.

This way, we remove one bit of coupling between Buildkite and this lane.

mokagio committed:
Add required value for `NSMicrophoneUsageDescription` to Reader

Aims to address the App Store Connect upload issues:

> ITMS-90683: Missing purpose string in Info.plist - Your app’s code references one or more APIs that access sensitive user data, or the app has one or more entitlements that permit such access.
> The Info.plist file for the “Reader.app” bundle should contain a NSMicrophoneUsageDescription key with a user-facing purpose string explaining clearly and completely why your app needs the data.
> If you’re using external libraries or SDKs, they may reference APIs that require a purpose string.
> While your app might not use these APIs, a purpose string is still required.
> For details, visit: https://developer.apple.com/documentation/uikit/protecting_the_user_s_privacy/requesting_access_to_protected_resources.

mokagio committed:
Add required value for `NSCameraUsageDescription` to Reader

Aims to address the App Store Connect upload issues:

> ITMS-90683: Missing purpose string in Info.plist - Your app’s code references one or more APIs that access sensitive user data, or the app has one or more entitlements that permit such access.
> The Info.plist file for the “Reader.app” bundle should contain a NSCameraUsageDescription key with a user-facing purpose string explaining clearly and completely why your app needs the data.
> If you’re using external libraries or SDKs, they may reference APIs that require a purpose string.
> While your app might not use these APIs, a purpose string is still required.
> For details, visit: https://developer.apple.com/documentation/uikit/protecting_the_user_s_privacy/requesting_access_to_protected_resources.

mokagio committed:
Point Reachability to Automattic's fork

This addresses the App Store Connect upload issue:

> ITMS-90338: Non-public API usage - The app references non-public symbols in Frameworks/Keystone.framework/Keystone: _kReachabilityChangedNotification.
> If method names in your source code match the private Apple APIs listed above, altering your method names will help prevent this app from being flagged in future submissions.
> In addition, note that one or more of the above APIs may be located in a static library that was included with your app.
> If so, they must be removed.
> For further information, visit the Technical Support Information at http://developer.apple.com/support/technical/

mokagio committed:
Prefix `bodyTextColor` and `titleTextColor` (unsed) with `wp`

This is to avoid a clash with Apple's private APIs in the dependency
setup with Keystone.

> ITMS-90338: Non-public API usage - The app references non-public selectors in Frameworks/Keystone.framework/Keystone: setBodyTextColor:, setTitleTextColor:.
> If method names in your source code match the private Apple APIs listed above, altering your method names will help prevent this app from being flagged in future submissions.
> In addition, note that one or more of the above APIs may be located in a static library that was included with your app.
> If so, they must be removed.
> For further information, visit the Technical Support Information at http://developer.apple.com/support/technical/

mokagio commented on Modules/Package.swift:
```diff
@@ -42,7 +42,7 @@ let package = Package(
         .package(url: "https://github.com/squarefrog/UIDeviceIdentifier", from: "2.3.0"),
         .package(url: "https://github.com/SVProgressHUD/SVProgressHUD", from: "2.3.1"),
-        .package(url: "https://github.com/tonymillion/Reachability", from: "3.7.5"),
+        .package(url: "https://github.com/Automattic/Reachability", branch: "framework-support-via-spm"),
```
Addresses:

> ITMS-90338: Non-public API usage - The app references non-public symbols in Frameworks/Keystone.framework/Keystone: _kReachabilityChangedNotification. If method names in your source code match the private Apple APIs listed above, altering your method names will help prevent this app from being flagged in future submissions. In addition, note that one or more of the above APIs may be located in a static library that was included with your app. If so, they must be removed. For further information, visit the Technical Support Information at http://developer.apple.com/support/technical/

Notice that this required also updating the notification name from `kReachability...` to `kTMReachability...` (TM stands for Tony Million, the original author).

While we shouldn't fix what is not broken and this minor update/fork should suffice, we might want to update from Reachability to a more modern implementation as some point.

crazytonyli replied:
I think a less drastic change is adding [the privacy info file](https://github.com/tonymillion/Reachability/blob/master/Framework/PrivacyInfo.xcprivacy) to SPM. The original repo added it to the podspec, but not SPM.

mokagio commented on Modules/Sources/WordPressUI/FancyAlert/FancyAlertView.swift:
```diff
@@ -62,7 +62,8 @@ open class FancyAlertView: UIView {
     ///
-    @objc public dynamic var titleTextColor: UIColor? {
+    /// Notice the "wp" prefix to avoid clashing with Apple's private API.
+    @objc public dynamic var wpTitleTextColor: UIColor? {
```
Notice that this is unused in the apps. Happy to remove it in a follow up PR to avoid future confusion.

mokagio commented on Modules/Sources/WordPressUI/FancyAlert/FancyAlertView.swift:
```diff
@@ -73,7 +74,8 @@ open class FancyAlertView: UIView {
+    /// Notice the "wp" prefix to avoid clashing with Apple's private API.
+    @objc public dynamic var wpBodyTextColor: UIColor? {
         get {
             return bodyLabel.textColor
```
FWIW, I tried accessing `bodyLabel.textColor` directly, but it's not possible—which is obviously why this wrapper exists

<img width="1384" alt="image" src="https://github.com/user-attachments/assets/02ac7567-fe37-4a17-a451-8a6fb2fd414d" />

mokagio committed:
Update Fastlane to latest version

mokagio committed:
Add `ITSAppUsesNonExemptEncryption = NO` to Reader

mokagio committed:
Disable distribution groups to attempt fixing TestFlight issue

See how this build fails
https://buildkite.com/automattic/wordpress-ios/builds/27323/steps?sid=01963655-580b-44c3-846f-ef84df40eb86

"Builds cannot be assigned to this internal group. - Cannot add internal group to a build."

However, on TestFlight itself the build is successfully part of the
group and distributed.

mokagio commented on Sources/Reader/Resources/LaunchScreen.storyboard:
```diff
@@ -0,0 +1,30 @@
+            <color white="1" alpha="1" colorSpace="custom" customColorSpace="genericGamma22GrayColorSpace"/>
+        </systemColor>
+    </resources>
+</document>
```
Addresses 

> [Application Loader Error Output]: [ContentDelivery.Uploader.6000033F8180] Validation failed (409) Invalid bundle. Apps that support Multitasking on iPad must provide the app’s launch screen using an Xcode storyboard, or using UILaunchScreen if the app’s MinimumOSVersion is 14 or higher. Verify that the UILaunchStoryboardName key is included in your com.automattic.reader bundle if you’re using a storyboard. For details, see: https://developer.apple.com/documentation/bundleresources/information_property_list/uilaunchstoryboardname (ID: 644f4981-9ca7-4cce-a458-309734a455e0)

See
https://buildkite.com/automattic/wordpress-ios/builds/27266/steps?sid=0196236d-400e-4be1-9cf8-399f2ca6b45c

mokagio commented on WordPress/WordPress.xcodeproj/project.pbxproj:
```diff
@@ -6678,6 +6678,12 @@
 				GCC_WARN_UNINITIALIZED_AUTOS = YES_AGGRESSIVE;
 				GENERATE_INFOPLIST_FILE = YES;
 				INFOPLIST_FILE = ../Sources/Reader/Info.plist;
+				INFOPLIST_KEY_ITSAppUsesNonExemptEncryption = NO;
+				INFOPLIST_KEY_NSCameraUsageDescription = "To take photos or videos to use in your posts.";
+				INFOPLIST_KEY_NSMicrophoneUsageDescription = "Enable microphone access to record sound in your videos.";
+				INFOPLIST_KEY_UILaunchStoryboardName = LaunchScreen;
+				INFOPLIST_KEY_UISupportedInterfaceOrientations = "UIInterfaceOrientationLandscapeLeft UIInterfaceOrientationLandscapeRight UIInterfaceOrientationPortrait";
+				INFOPLIST_KEY_UISupportedInterfaceOrientations_iPad = "UIInterfaceOrientationLandscapeLeft UIInterfaceOrientationLandscapeRight UIInterfaceOrientationPortrait UIInterfaceOrientationPortraitUpsideDown";
```
Address

> [Application Loader Error Output]: [ContentDelivery.Uploader.6000033F8180] Validation failed (409) Invalid bundle. Apps that support Multitasking on iPad must provide the app’s launch screen using an Xcode storyboard, or using UILaunchScreen if the app’s MinimumOSVersion is 14 or higher. Verify that the UILaunchStoryboardName key is included in your com.automattic.reader bundle if you’re using a storyboard. For details, see: https://developer.apple.com/documentation/bundleresources/information_property_list/uilaunchstoryboardname (ID: 644f4981-9ca7-4cce-a458-309734a455e0)

> ITMS-90683: Missing purpose string in Info.plist - Your app’s code references one or more APIs that access sensitive user data, or the app has one or more entitlements that permit such access. The Info.plist file for the “Reader.app” bundle should contain a NSMicrophoneUsageDescription key with a user-facing purpose string explaining clearly and completely why your app needs the data. If you’re using external libraries or SDKs, they may reference APIs that require a purpose string. While your app might not use these APIs, a purpose string is still required. For details, visit: https://developer.apple.com/documentation/uikit/protecting_the_user_s_privacy/requesting_access_to_protected_resources.

> ITMS-90683: Missing purpose string in Info.plist - Your app’s code references one or more APIs that access sensitive user data, or the app has one or more entitlements that permit such access. The Info.plist file for the “Reader.app” bundle should contain a NSCameraUsageDescription key with a user-facing purpose string explaining clearly and completely why your app needs the data. If you’re using external libraries or SDKs, they may reference APIs that require a purpose string. While your app might not use these APIs, a purpose string is still required. For details, visit: https://developer.apple.com/documentation/uikit/protecting_the_user_s_privacy/requesting_access_to_protected_resources.
